src/main/java/com/github/packageurl/PackageURL.java

```diff
@@ -460,7 +460,7 @@ private static String uriEncode(String source, Charset charset) {
     }
 
     private static boolean isUnreserved(int c) {
-        return (isAlpha(c) || isDigit(c) || '-' == c || '.' == c || '_' == c || '~' == c);
+        return (isAlpha(c) || isDigit(c) || '-' == c || '.' == c || '_' == c || '~' == c || ':' == c || '/' == c);
```

Comment:
There should be rather a different set for each PURL component:

- `namespace` does not need to encode neither `:` nor `/`.
- `name` does not need to encode `:`, but it needs to encode `/`, which is used to separate it from `namespace`.
- `qualifiers`, probably does not need to encode `:`, `/` or `?`.

Reply:
I will add a param to the `percentEncode` to take a `String` of characters to allow.

Reply:
Really, there are a lot more characters potentially allowed by the RFC than we list here.

I am of the opinion right now to add just enough in order to pass the test suite.

The `'?'` is not allowed unencoded per the purl spec itself.